src/vos/ilog.c

```diff
@@ -0,0 +1,1474 @@
+	.to_rec_free		= ilog_rec_free,
+	.to_rec_fetch		= ilog_rec_fetch,
+	.to_rec_update		= ilog_rec_update,
+};
+
+int
+ilog_init(void)
+{
+	int	rc;
+
+	rc = dbtree_class_register(VOS_BTR_ILOG, 0, &ilog_btr_ops);
+	if (rc != 0)
+		D_ERROR("Failed to register incarnation log btree class: %s\n",
+			d_errstr(rc));
+
+	return rc;
+}
+
+/** Called when we know a txn is needed.  Subsequent calls are a noop. */
+static inline int
+ilog_tx_begin(struct ilog_context *lctx)
+{
+	int	rc = 0;
+
+	if (lctx->ic_in_txn)
+		return 0;
+
+	rc = vos_tx_begin(&lctx->ic_umm);
+	if (rc != 0)
+		return rc;
+
+	lctx->ic_in_txn = true;
+	return 0;
+}
+
+/** Only invokes transaction end if we've started a txn */
+static inline int
+ilog_tx_end(struct ilog_context *lctx, int rc)
+{
+	if (!lctx->ic_in_txn)
+		return rc;
+
+	lctx->ic_in_txn = false;
+	return vos_tx_end(&lctx->ic_umm, rc);
+}
+
+static inline bool
+ilog_empty(struct ilog_root *root)
+{
+	return !root->lr_tree.it_embedded &&
+		root->lr_tree.it_root == UMOFF_NULL;
+}
+
+#define ILOG_MAGIC 0xdeadbaad
+
+static void
+ilog_addref(struct ilog_context *lctx)
+{
+	lctx->ic_ref++;
+}
+
+static void
+ilog_decref(struct ilog_context *lctx)
+{
+	lctx->ic_ref--;
+	if (lctx->ic_ref == 0)
+		D_FREE(lctx);
+}
+
+static int
+ilog_ctx_create(struct umem_instance *umm, struct ilog_root *root,
+		    struct ilog_context **lctxp)
+{
+	D_ALLOC_PTR(*lctxp);
+	if (*lctxp == NULL) {
+		D_ERROR("Could not allocate memory for open incarnation log\n");
+		return -DER_NOMEM;
+	}
+
+	(*lctxp)->ic_root = root;
+	(*lctxp)->ic_root_off = umem_ptr2off(umm, root);
+	(*lctxp)->ic_umm = *umm;
+	ilog_addref(*lctxp);
+	return 0;
+}
+
+static daos_handle_t
+ilog_lctx2hdl(struct ilog_context *lctx)
+{
+	daos_handle_t	hdl;
+
+	hdl.cookie = (uint64_t)lctx;
+
+	return hdl;
+}
+
+static struct ilog_context *
+ilog_hdl2lctx(daos_handle_t hdl)
+{
+	struct ilog_context	*lctx;
+
+	if (daos_handle_is_inval(hdl))
+		return NULL;
+
+	lctx = (struct ilog_context *)hdl.cookie;
+
+	if (lctx->ic_root->lr_magic != ILOG_MAGIC)
+		return NULL;
+
+	return lctx;
+}
+
+static int
+ilog_ptr_set_full(struct ilog_context *lctx, void *dest, const void *src,
+		  size_t len)
+{
+	int	rc = 0;
+
+	rc = ilog_tx_begin(lctx);
+	if (rc != 0) {
+		D_ERROR("Failed to start PMDK transaction: rc = %s\n",
+			d_errstr(rc));
+		goto done;
+	}
+
+	rc = umem_tx_add_ptr(&lctx->ic_umm, dest, len);
+	if (rc != 0) {
+		D_ERROR("Failed to add to undo log\n");
+		goto done;
+	}
+
+	memcpy(dest, src, len);
+done:
+	return rc;
+}
+
+#define ilog_ptr_set(lctx, dest, src)	\
+	ilog_ptr_set_full(lctx, dest, src, sizeof(*(src)))
+
+#define ilog_entry_set
+
+int
+ilog_create(struct umem_instance *umm, struct ilog_df *root)
+{
+	struct ilog_context	lctx = {
+		.ic_root = (struct ilog_root *)root,
+		.ic_umm = *umm,
+		.ic_ref = 0,
+		.ic_in_txn = 0,
+	};
+	struct ilog_root	tmp = {0};
+	int			rc = 0;
+
+	tmp.lr_magic = ILOG_MAGIC;
+	rc = ilog_ptr_set(&lctx, root, &tmp);
+
+	rc = ilog_tx_end(&lctx, rc);
```

Comment:
The issue is I don't start a transaction unless something has been updated.  Since the actual update is conditional, when we start a transaction, we set a flag.   ilog_tx_end checks that flag and only calls umem_tx_end if the flag is set.